File: src/data/courses/dsa/stacks/stack-advanced/max-stack/code/python/max_stack.py
# ============================================================
# Max Stack
# ============================================================

class MaxStack:
    def __init__(self):
        self.main = []
        self.max_stack = []

    def push(self, val: int) -> None:
        self.main.append(val)
        self.max_stack.append(
            val if not self.max_stack else max(val, self.max_stack[-1])
        )

    def pop(self) -> None:
        self.main.pop()
        self.max_stack.pop()

    def top(self) -> int:
        return self.main[-1]

    def get_max(self) -> int:
        return self.max_stack[-1]

File: src/data/courses/dsa/stacks/stack-advanced/max-stack/code/cpp/max_stack.cpp
// ============================================================
// Max Stack
// ============================================================

#include <stack>
#include <algorithm>
using namespace std;

class MaxStack {
    stack<int> mainStack;
    stack<int> maxStack;
public:
    void push(int val) {
        mainStack.push(val);
        maxStack.push(maxStack.empty() ? val : max(val, maxStack.top()));
    }
    void pop() { mainStack.pop(); maxStack.pop(); }
    int top() { return mainStack.top(); }
    int getMax() { return maxStack.top(); }
};
